core/services/directrequest/delegate.go

```diff
@@ -101,25 +125,99 @@ func (d listener) HandleLog(lb log.Broadcast) {
 		return
 	}
 
-	// TODO: Logic to handle log will go here
+	logJobSpecID := lb.RawLog().Topics[1]
+	if logJobSpecID == (common.Hash{}) || logJobSpecID != l.onChainJobSpecID {
+		logger.Debugw("Skipping Run for Log with wrong Job ID", "logJobSpecID", logJobSpecID, "actualJobID", l.onChainJobSpecID)
+		return
+	}
 
-	err = lb.MarkConsumed()
-	if err != nil {
-		logger.Errorf("Error marking log as consumed: %v", err)
+	log := lb.DecodedLog()
+	if log == nil || reflect.ValueOf(log).IsNil() {
+		logger.Error("HandleLog: ignoring nil value")
+		return
+	}
+
+	switch log := log.(type) {
+	case *oracle_wrapper.OracleOracleRequest:
+		l.handleOracleRequest(log)
+		err = lb.MarkConsumed()
+		if err != nil {
+			logger.Errorf("Error marking log as consumed: %v", err)
+		}
+	case *oracle_wrapper.OracleCancelOracleRequest:
+		l.handleCancelOracleRequest(log)
+		err = lb.MarkConsumed()
+		if err != nil {
+			logger.Errorf("Error marking log as consumed: %v", err)
+		}
+
+	default:
+		logger.Warnf("unexpected log type %T", log)
+	}
+}
+
+func oracleRequestToMap(request *oracle_wrapper.OracleOracleRequest) map[string]interface{} {
+	result := make(map[string]interface{})
+	result["specId"] = fmt.Sprintf("0x%x", request.SpecId)
+	result["requester"] = request.Requester.Hex()
+	result["requestId"] = formatRequestId(request.RequestId)
+	result["payment"] = fmt.Sprintf("%v", request.Payment)
+	result["callbackAddr"] = request.CallbackAddr.Hex()
+	result["callbackFunctionId"] = fmt.Sprintf("0x%x", request.CallbackFunctionId)
+	result["cancelExpiration"] = fmt.Sprintf("%v", request.CancelExpiration)
+	result["dataVersion"] = fmt.Sprintf("%v", request.DataVersion)
+	result["data"] = fmt.Sprintf("0x%x", request.Data)
+	return result
+}
+
+func (l *listener) handleOracleRequest(request *oracle_wrapper.OracleOracleRequest) {
+	meta := make(map[string]interface{})
+	meta["oracleRequest"] = oracleRequestToMap(request)
+
+	logger := logger.CreateLogger(logger.Default.With(
+		"jobName", l.spec.JobName,
+		"jobID", l.spec.JobID,
+	))
+
+	go func() {
+		runCloserChannel := make(chan struct{})
+		runCloserChannelIf, loaded := l.runs.LoadOrStore(formatRequestId(request.RequestId), runCloserChannel)
+		if loaded {
+			runCloserChannel, _ = runCloserChannelIf.(chan struct{})
+		}
+		ctx, cancel := utils.CombinedContext(runCloserChannel, context.Background())
```

Comment:
That gets a little ugly because you can't really use the `defer cancel()` pattern, because the context will be shared amongst multiple go routines. I think you really want a single context for each `Execute...`

Reply:
not sure I follow - the context is thread safe?